src/libaktualizr/primary/sotauptaneclient.h

```diff
@@ -30,9 +31,26 @@
 class SotaUptaneClient {
  public:
   SotaUptaneClient(Config &config_in, const std::shared_ptr<INvStorage> &storage_in,
-                   std::shared_ptr<HttpInterface> http_in = nullptr,
-                   std::shared_ptr<event::Channel> events_channel_in = nullptr);
-  ~SotaUptaneClient();
+                   std::shared_ptr<HttpInterface> http_in, std::shared_ptr<event::Channel> events_channel_in)
+      : config(config_in),
+        uptane_manifest(config, storage_in),
+        storage(storage_in),
+        http(std::move(http_in)),
+        package_manager(PackageManagerFactory::makePackageManager(config.pacman, config.bootloader, storage, http)),
+        uptane_fetcher(new Uptane::Fetcher(config, http)),
+        report_queue(new ReportQueue(config, http)),
+        events_channel(std::move(events_channel_in)) {}
+
+  SotaUptaneClient(Config &config_in, const std::shared_ptr<INvStorage> &storage_in,
+                   std::shared_ptr<event::Channel> events_channel_in)
+      : SotaUptaneClient(config_in, storage_in, std::make_shared<HttpClient>(), std::move(events_channel_in)) {}
+
+  SotaUptaneClient(Config &config_in, const std::shared_ptr<INvStorage> &storage_in,
+                   std::shared_ptr<HttpInterface> http_in)
+      : SotaUptaneClient(config_in, storage_in, std::move(http_in), nullptr) {}
+
+  SotaUptaneClient(Config &config_in, const std::shared_ptr<INvStorage> &storage_in)
+      : SotaUptaneClient(config_in, storage_in, std::make_shared<HttpClient>(), nullptr) {}
```

Comment:
What makes having the multiple constructors better than using default arguments?

Reply:
Default arguments should be checked at runtime vs compile time for constructors, that makes default constructors a safer option IMO.